### PR TITLE
build-in: add cot build-in function

### PR DESCRIPTION
**Pull request by dsh0416**

### CODE REVIEW

**Comment by coocood** [comment]
@dsh0416 
Please fix CI.

**Comment by dsh0416** [comment]
@coocood CI fixed

File: expression/builtin_math.go [hunk 1]
```diff
@@ -831,7 +831,26 @@ type builtinCotSig struct {
...
+	if arg.IsNull() {
+		return d, nil
+	}
+	sc := b.ctx.GetSessionVars().StmtCtx
+	x, err := arg.ToFloat64(sc)
+	if err != nil {
+		return d, errors.Trace(err)
+	}
+	result := (1.0 / math.Tan(x))
+	if math.IsInf(result, 0) == true {
```
Comment (coocood) [inline expression/builtin_math.go]: MySQL returns error when argument is 0.

Comment (dsh0416) [inline expression/builtin_math.go]: MySQL should return NULL when argument is 0 due to [doc](https://dev.mysql.com/doc/refman/5.7/en/mathematical-functions.html#function_cot). But it seems that I met some problems with typeinferer.

File: expression/builtin_math.go [hunk 2]
```diff
@@ -831,7 +831,26 @@ type builtinCotSig struct {
...
+	arg := args[0]
+	if arg.IsNull() {
+		return d, nil
+	}
+	sc := b.ctx.GetSessionVars().StmtCtx
+	x, err := arg.ToFloat64(sc)
+	if err != nil {
+		return d, errors.Trace(err)
+	}
+	result := (1.0 / math.Tan(x))
```
Comment (XuHuaiyu) [inline expression/builtin_math.go]: The outer parenthesis maybe useless.

File: expression/builtin_math.go [hunk 3]
```diff
@@ -831,7 +831,26 @@ type builtinCotSig struct {
 
 // See https://dev.mysql.com/doc/refman/5.7/en/mathematical-functions.html#function_cot
 func (b *builtinCotSig) eval(row []types.Datum) (d types.Datum, err error) {
-	return d, errFunctionNotExists.GenByArgs("cot")
+	args, err := b.evalArgs(row)
+	if err != nil {
+		return types.Datum{}, errors.Trace(err)
```
Comment (XuHuaiyu) [inline expression/builtin_math.go]: return d, errors.Trace(err)

File: expression/builtin_math.go [hunk 4]
```diff
@@ -831,7 +831,26 @@ type builtinCotSig struct {
...
+	sc := b.ctx.GetSessionVars().StmtCtx
+	x, err := arg.ToFloat64(sc)
+	if err != nil {
+		return d, errors.Trace(err)
+	}
+	result := (1.0 / math.Tan(x))
+	if math.IsInf(result, 0) == true {
+		d.SetNull()
+	} else {
+		d.SetFloat64(1.0 / math.Tan(x))
```
Comment (XuHuaiyu) [inline expression/builtin_math.go]: why not d.SetFloat64(result)?

File: expression/builtin_math.go (hunk 1 above)
Comment (XuHuaiyu) [inline expression/builtin_math.go]: `==true` is useless.

File: expression/builtin_math.go (hunk 4 above)
Comment (dsh0416) [inline expression/builtin_math.go]: Based on MySQL doc, Inf should be returned as Null. I'm not quite sure about how the interface built in tidb works. It seems to be easy if it doesn't matter.

**Comment by shenli** [comment]
@dsh0416 Any update？Please resolve the conflicts.

**Comment by XuHuaiyu** [comment]
@dsh0416 
hi, any updates?

**Comment by coocood** [comment]
@dsh0416 
I'm sorry, this function has been implemented already.

**Pull request closed without merging**